File: tools/revit-sim/src/revit_sim/placement.py
"""Shared placement math (Part G projection primitive specialized to offset-along-wall).

Pinned against packages/contracts/fixtures/placement/manifest.json — the C# twin is
ChapterHub.Core.Placement; the Phase 1 acceptance requires both to agree to 1e-6 mm.
"""

from __future__ import annotations

import math

Pt2 = tuple[float, float]


def _unit(start: Pt2, end: Pt2) -> Pt2:
    dx, dy = end[0] - start[0], end[1] - start[1]
    length = math.hypot(dx, dy)
    if length == 0:
        raise ValueError("zero-length wall")
    return dx / length, dy / length


def centerline_point(start: Pt2, end: Pt2, offset_mm: float) -> Pt2:
    ux, uy = _unit(start, end)
    return start[0] + ux * offset_mm, start[1] + uy * offset_mm


def face_point(start: Pt2, end: Pt2, thickness_mm: float, offset_mm: float, side: str) -> Pt2:
    """side: 'left' | 'right' of the start->end direction; left = +90deg CCW."""
    cx, cy = centerline_point(start, end, offset_mm)
    ux, uy = _unit(start, end)
    nx, ny = -uy, ux  # left normal
    if side == "right":
        nx, ny = -nx, -ny
    elif side != "left":
        raise ValueError(f"unknown side {side!r}")
    half = thickness_mm / 2.0
    return cx + nx * half, cy + ny * half


def place(
    kind: str,
    start: Pt2,
    end: Pt2,
    thickness_mm: float,
    offset_mm: float,
    z_mm: float,
) -> tuple[float, float, float]:
    """kind: 'centerline' | 'face_left' | 'face_right' (the fixture vocabulary)."""
    if kind == "centerline":
        x, y = centerline_point(start, end, offset_mm)
    elif kind in ("face_left", "face_right"):
        x, y = face_point(start, end, thickness_mm, offset_mm, kind.removeprefix("face_"))
    else:
        raise ValueError(f"unknown placement kind {kind!r}")
    return x, y, z_mm

File: plugin/ChapterHub.Core/Placement.cs
using ChapterHub.Core.Contracts;

namespace ChapterHub.Core;

/// <summary>
/// Shared placement math (Part G projection primitive specialized to offset-along-wall),
/// in millimeters. Pinned against packages/contracts/fixtures/placement/manifest.json;
/// the Python twin is revit_sim.placement — Phase 1 acceptance requires both to agree
/// to 1e-6 mm. The Addin's op handlers convert the result with UnitConversion.MmToFt.
/// </summary>
public static class Placement
{
    public readonly record struct Point3(double X, double Y, double Z);

    /// <param name="kind">"centerline" | "face_left" | "face_right" (the fixture vocabulary);
    /// left = +90° CCW of the start→end direction.</param>
    public static Point3 Place(
        string kind, Pt2 start, Pt2 end, double thicknessMm, double offsetMm, double zMm)
    {
        var (ux, uy) = Unit(start, end);
        var cx = start.X + ux * offsetMm;
        var cy = start.Y + uy * offsetMm;
        switch (kind)
        {
            case "centerline":
                return new Point3(cx, cy, zMm);
            case "face_left":
            case "face_right":
            {
                var (nx, ny) = (-uy, ux); // left normal
                if (kind == "face_right") (nx, ny) = (-nx, -ny);
                var half = thicknessMm / 2.0;
                return new Point3(cx + nx * half, cy + ny * half, zMm);
            }
            default:
                throw new ArgumentException($"unknown placement kind '{kind}'", nameof(kind));
        }
    }

    private static (double Ux, double Uy) Unit(Pt2 start, Pt2 end)
    {
        var dx = end.X - start.X;
        var dy = end.Y - start.Y;
        var length = Math.Sqrt(dx * dx + dy * dy);
        if (length == 0) throw new ArgumentException("zero-length wall");
        return (dx / length, dy / length);
    }
}
